In [42]:
import numpy as np,pandas as pd,matplotlib.pyplot as plt,seaborn as sns

##Loading the data and getting basic idea about it

In [43]:
df=pd.read_csv('customer_churn.csv')
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

In [44]:
df.shape

(7043, 21)

In [45]:
df.head()

customerID  gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  7590-VHVEG  Female              0     Yes         No       1           No   
1  5575-GNVDE    Male              0      No         No      34          Yes   
2  3668-QPYBK    Male              0      No         No       2          Yes   
3  7795-CFOCW    Male              0      No         No      45           No   
4  9237-HQITU  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity  ... DeviceProtection  \
0  No phone service             DSL             No  ...               No   
1                No             DSL            Yes  ...              Yes   
2                No             DSL            Yes  ...               No   
3  No phone service             DSL            Yes  ...              Yes   
4                No     Fiber optic             No  ...               No   

  TechSupport StreamingTV StreamingMovies        Contract PaperlessBilling  \
0          No          No              No  Month-to-month              Yes   
1          No          No              No        One year               No   
2          No          No              No  Month-to-month              Yes   
3         Yes          No              No        One year               No   
4          No          No              No  Month-to-month              Yes   

               PaymentMethod MonthlyCharges  TotalCharges Churn  
0           Electronic check          29.85         29.85    No  
1               Mailed check          56.95        1889.5    No  
2               Mailed check          53.85        108.15   Yes  
3  Bank transfer (automatic)          42.30       1840.75    No  
4           Electronic check          70.70        151.65   Yes  

[5 rows x 21 columns]

The last column 'Churn' is the target column. Churn means whether the customer is leaving the company or not. Goal of this handson is to predict whether a customer will port his sim to a new company (churn) or not

In [46]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7043 entries, 0 to 7042
Data columns (total 21 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   customerID        7043 non-null   object 
 1   gender            7043 non-null   object 
 2   SeniorCitizen     7043 non-null   int64  
 3   Partner           7043 non-null   object 
 4   Dependents        7043 non-null   object 
 5   tenure            7043 non-null   int64  
 6   PhoneService      7043 non-null   object 
 7   MultipleLines     7043 non-null   object 
 8   InternetService   7043 non-null   object 
 9   OnlineSecurity    7043 non-null   object 
 10  OnlineBackup      7043 non-null   object 
 11  DeviceProtection  7043 non-null   object 
 12  TechSupport       7043 non-null   object 
 13  StreamingTV       7043 non-null   object 
 14  StreamingMovies   7043 non-null   object 
 15  Contract          7043 non-null   object 
 16  PaperlessBilling  7043 non-null   object 


In [47]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges         object
Churn                object
dtype: object

In [48]:
df.describe()

SeniorCitizen       tenure  MonthlyCharges
count    7043.000000  7043.000000     7043.000000
mean        0.162147    32.371149       64.761692
std         0.368612    24.559481       30.090047
min         0.000000     0.000000       18.250000
25%         0.000000     9.000000       35.500000
50%         0.000000    29.000000       70.350000
75%         0.000000    55.000000       89.850000
max         1.000000    72.000000      118.750000

In [49]:
df.describe(include='O')    #o will give statistical summary for on;y object type columns
#all will give for all columns

customerID gender Partner Dependents PhoneService MultipleLines  \
count         7043   7043    7043       7043         7043          7043   
unique        7043      2       2          2            2             3   
top     7590-VHVEG   Male      No         No          Yes            No   
freq             1   3555    3641       4933         6361          3390   

       InternetService OnlineSecurity OnlineBackup DeviceProtection  \
count             7043           7043         7043             7043   
unique               3              3            3                3   
top        Fiber optic             No           No               No   
freq              3096           3498         3088             3095   

       TechSupport StreamingTV StreamingMovies        Contract  \
count         7043        7043            7043            7043   
unique           3           3               3               3   
top             No          No              No  Month-to-month   
freq          3473        2810            2785            3875   

       PaperlessBilling     PaymentMethod TotalCharges Churn  
count              7043              7043         7043  7043  
unique                2                 4         6531     2  
top                 Yes  Electronic check                 No  
freq               4171              2365           11  5174

##Performing EDA
Data can be manipulated by 2 things:
1. Null values
2. Duplicates

We need to treat these values before performing any analysis or else they may affect accuracy of prediction

- Null values can either be replaced or removed
- Duplicates can be dropped


In [50]:
df.isnull().sum()

customerID          0
gender              0
SeniorCitizen       0
Partner             0
Dependents          0
tenure              0
PhoneService        0
MultipleLines       0
InternetService     0
OnlineSecurity      0
OnlineBackup        0
DeviceProtection    0
TechSupport         0
StreamingTV         0
StreamingMovies     0
Contract            0
PaperlessBilling    0
PaymentMethod       0
MonthlyCharges      0
TotalCharges        0
Churn               0
dtype: int64

'TotalCharges' column contains float number however, the datatype of that column is object.

In [51]:
df['TotalCharges'].value_counts()

          11
20.2      11
19.75      9
20.05      8
19.9       8
          ..
6849.4     1
692.35     1
130.15     1
3211.9     1
6844.5     1
Name: TotalCharges, Length: 6531, dtype: int64

From above we can observe that there are 11 empty values which are stored as object data type which is resulting in incorrect assignment of data type to the rest of numerical column.

These empty values are not counted in null values

In [52]:
df[df['TotalCharges']==' ']

customerID  gender  SeniorCitizen Partner Dependents  tenure  \
488   4472-LVYGI  Female              0     Yes        Yes       0   
753   3115-CZMZD    Male              0      No        Yes       0   
936   5709-LVOEQ  Female              0     Yes        Yes       0   
1082  4367-NUYAO    Male              0     Yes        Yes       0   
1340  1371-DWPAZ  Female              0     Yes        Yes       0   
3331  7644-OMVMY    Male              0     Yes        Yes       0   
3826  3213-VVOLG    Male              0     Yes        Yes       0   
4380  2520-SGTTA  Female              0     Yes        Yes       0   
5218  2923-ARZLG    Male              0     Yes        Yes       0   
6670  4075-WKNIU  Female              0     Yes        Yes       0   
6754  2775-SEFEE    Male              0      No        Yes       0   

     PhoneService     MultipleLines InternetService       OnlineSecurity  ...  \
488            No  No phone service             DSL                  Yes  ...   
753           Yes                No              No  No internet service  ...   
936           Yes                No             DSL                  Yes  ...   
1082          Yes               Yes              No  No internet service  ...   
1340           No  No phone service             DSL                  Yes  ...   
3331          Yes                No              No  No internet service  ...   
3826          Yes               Yes              No  No internet service  ...   
4380          Yes                No              No  No internet service  ...   
5218          Yes                No              No  No internet service  ...   
6670          Yes               Yes             DSL                   No  ...   
6754          Yes               Yes             DSL                  Yes  ...   

         DeviceProtection          TechSupport          StreamingTV  \
488                   Yes                  Yes                  Yes   
753   No internet service  No internet service  No internet service   
936                   Yes                   No                  Yes   
1082  No internet service  No internet service  No internet service   
1340                  Yes                  Yes                  Yes   
3331  No internet service  No internet service  No internet service   
3826  No internet service  No internet service  No internet service   
4380  No internet service  No internet service  No internet service   
5218  No internet service  No internet service  No internet service   
6670                  Yes                  Yes                  Yes   
6754                   No                  Yes                   No   

          StreamingMovies  Contract PaperlessBilling  \
488                    No  Two year              Yes   
753   No internet service  Two year               No   
936                   Yes  Two year               No   
1082  No internet service  Two year               No   
1340                   No  Two year               No   
3331  No internet service  Two year               No   
3826  No internet service  Two year               No   
4380  No internet service  Two year               No   
5218  No internet service  One year              Yes   
6670                   No  Two year               No   
6754                   No  Two year              Yes   

                  PaymentMethod MonthlyCharges  TotalCharges Churn  
488   Bank transfer (automatic)          52.55                  No  
753                Mailed check          20.25                  No  
936                Mailed check          80.85                  No  
1082               Mailed check          25.75                  No  
1340    Credit card (automatic)          56.05                  No  
3331               Mailed check          19.85                  No  
3826               Mailed check          25.35                  No  
4380               Mailed check          20.00                  No  
5218               Mailed check          19.7

Above 11 records have no values in the TotalCharges column. We can replace these values with mean or NaN

Here as there are only 11 such records which is very few relative to total records, we can directly drop these records

In [53]:
df['TotalCharges']=pd.to_numeric(df['TotalCharges'],errors='coerce')
#this will convert all empty values to NaN(by coerce) and then change dtype to numeric(to_num)
#replacing and converting to float is done by above single command

In [54]:
df.isnull().sum()

customerID           0
gender               0
SeniorCitizen        0
Partner              0
Dependents           0
tenure               0
PhoneService         0
MultipleLines        0
InternetService      0
OnlineSecurity       0
OnlineBackup         0
DeviceProtection     0
TechSupport          0
StreamingTV          0
StreamingMovies      0
Contract             0
PaperlessBilling     0
PaymentMethod        0
MonthlyCharges       0
TotalCharges        11
Churn                0
dtype: int64

Now 11 records are detected while checking for null

In [55]:
df.dtypes

customerID           object
gender               object
SeniorCitizen         int64
Partner              object
Dependents           object
tenure                int64
PhoneService         object
MultipleLines        object
InternetService      object
OnlineSecurity       object
OnlineBackup         object
DeviceProtection     object
TechSupport          object
StreamingTV          object
StreamingMovies      object
Contract             object
PaperlessBilling     object
PaymentMethod        object
MonthlyCharges      float64
TotalCharges        float64
Churn                object
dtype: object

Above list confirms that the data type of Total Charges column is converted to float and updated in list successfully

In [56]:
df=df.dropna()

In [57]:
df.isnull().sum().sum()

0

All null values are removed

In [58]:
df.duplicated().sum()

0

Customer ID column is irrelevant here as it doesn't influence any decision of predicting churn data. This column can be dropped

In [59]:
df.drop(columns=['customerID'],inplace=True)

In [60]:
df.head()

gender  SeniorCitizen Partner Dependents  tenure PhoneService  \
0  Female              0     Yes         No       1           No   
1    Male              0      No         No      34          Yes   
2    Male              0      No         No       2          Yes   
3    Male              0      No         No      45           No   
4  Female              0      No         No       2          Yes   

      MultipleLines InternetService OnlineSecurity OnlineBackup  \
0  No phone service             DSL             No          Yes   
1                No             DSL            Yes           No   
2                No             DSL            Yes          Yes   
3  No phone service             DSL            Yes           No   
4                No     Fiber optic             No           No   

  DeviceProtection TechSupport StreamingTV StreamingMovies        Contract  \
0               No          No          No              No  Month-to-month   
1              Yes          No          No              No        One year   
2               No          No          No              No  Month-to-month   
3              Yes         Yes          No              No        One year   
4               No          No          No              No  Month-to-month   

  PaperlessBilling              PaymentMethod  MonthlyCharges  TotalCharges  \
0              Yes           Electronic check           29.85         29.85   
1               No               Mailed check           56.95       1889.50   
2              Yes               Mailed check           53.85        108.15   
3               No  Bank transfer (automatic)           42.30       1840.75   
4              Yes           Electronic check           70.70        151.65   

  Churn  
0    No  
1    No  
2   Yes  
3    No  
4   Yes

For tree based problems, outliers need not be removed as they don't affect the result to a great extent. Outliers are expected to be removed in case of linear and logistic regression

##Encoding with help of Label Encoding

In [61]:
col_list=[]
for i in df.columns:
    if (df[i].dtype=='object')& (i!='Churn'):
        col_list.append(i)

col_list

['gender',
 'Partner',
 'Dependents',
 'PhoneService',
 'MultipleLines',
 'InternetService',
 'OnlineSecurity',
 'OnlineBackup',
 'DeviceProtection',
 'TechSupport',
 'StreamingTV',
 'StreamingMovies',
 'Contract',
 'PaperlessBilling',
 'PaymentMethod']

Above 'col_list' contains the list of all object type columns excluding the target column Churn. Label encoding can be directly performed on this col_list

In [62]:
from sklearn.preprocessing import LabelEncoder
le=LabelEncoder()

In [63]:
for i in col_list:
    df[i]=le.fit_transform(df[i])

In [64]:
df.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0       1             0   
1       1              0        0           0      34             1   
2       1              0        0           0       2             1   
3       1              0        0           0      45             0   
4       0              0        0           0       2             1   

   MultipleLines  InternetService  OnlineSecurity  OnlineBackup  \
0              1                0               0             2   
1              0                0               2             0   
2              0                0               2             2   
3              1                0               2             0   
4              0                1               0             0   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 2            0            0                0         1   
2                 0            0            0                0         0   
3                 2            2            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges Churn  
0                 1              2           29.85         29.85    No  
1                 0              3           56.95       1889.50    No  
2                 1              3           53.85        108.15   Yes  
3                 0              0           42.30       1840.75    No  
4                 1              2           70.70        151.65   Yes

All the object columns are converted to numeric datatype after performing label encoding

##Preparing data for applying Random Forest
Dividing the dataset into dependant and independant columns

In [65]:
x=df.iloc[:,:-1] #features
y=df['Churn']    #target

In [66]:
x.head()

gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0       0              0        1           0       1             0   
1       1              0        0           0      34             1   
2       1              0        0           0       2             1   
3       1              0        0           0      45             0   
4       0              0        0           0       2             1   

   MultipleLines  InternetService  OnlineSecurity  OnlineBackup  \
0              1                0               0             2   
1              0                0               2             0   
2              0                0               2             2   
3              1                0               2             0   
4              0                1               0             0   

   DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                 0            0            0                0         0   
1                 2            0            0                0         1   
2                 0            0            0                0         0   
3                 2            2            0                0         1   
4                 0            0            0                0         0   

   PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges  
0                 1              2           29.85         29.85  
1                 0              3           56.95       1889.50  
2                 1              3           53.85        108.15  
3                 0              0           42.30       1840.75  
4                 1              2           70.70        151.65

In [67]:
y.head()

0     No
1     No
2    Yes
3     No
4    Yes
Name: Churn, dtype: object

Now divide the dataset into training and test set

In [68]:
from sklearn.model_selection import train_test_split

x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.2,random_state=0)    #usual split is 75,25. Train set should always be greater than test data

In [69]:
print(x_train.shape)
print(x_test.shape)
print(y_train.shape)
print(y_test.shape)

(5625, 19)
(1407, 19)
(5625,)
(1407,)


In [70]:
from sklearn.ensemble import RandomForestClassifier

classifier1=RandomForestClassifier(n_estimators=100)    #n_estimators is number of decision trees
classifier1.fit(x_train,y_train)

RandomForestClassifier()

In [71]:
y_pred=classifier1.predict(x_test)
y_pred  #prediction values given by model
#acutal values are stored in y_test

array(['No', 'No', 'No', ..., 'Yes', 'No', 'No'], dtype=object)

In [72]:
y_test

5561     No
5814     No
2645     No
3983    Yes
6438    Yes
       ... 
2757     No
5702    Yes
1662    Yes
2766     No
2918     No
Name: Churn, Length: 1407, dtype: object

In [73]:
from sklearn.metrics import accuracy_score
accuracy_score(y_pred,y_test)

0.7917555081734187

In [74]:
from sklearn.ensemble import RandomForestClassifier

classifier1=RandomForestClassifier(n_estimators=200)
classifier1.fit(x_train,y_train)

y_pred=classifier1.predict(x_test)

from sklearn.metrics import accuracy_score
accuracy_score(y_pred,y_test)

0.7938877043354655

In [77]:
from sklearn.ensemble import RandomForestClassifier

classifier1=RandomForestClassifier(n_estimators=300)
classifier1.fit(x_train,y_train)

y_pred=classifier1.predict(x_test)

from sklearn.metrics import accuracy_score
accuracy_score(y_pred,y_test)

0.7953091684434968

Accuracy is improving generally with increase in n-estimators(no. of decision trees)

In [78]:
n=[100,200,300,400,500,600,700,800,900,1000]

acc=[]
for i in n:
    model=RandomForestClassifier(n_estimators=i)
    model.fit(x_train,y_train)
    pred=model.predict(x_test)
    a=accuracy_score(pred,y_test)
    acc.append(a)

In [79]:
acc

[0.7945984363894811,
 0.7874911158493249,
 0.7924662402274343,
 0.7924662402274343,
 0.7896233120113717,
 0.7910447761194029,
 0.7896233120113717,
 0.7889125799573561,
 0.7931769722814499,
 0.7882018479033405]

In [80]:
d={'Number of trees':n, 'Accuracy':acc}
acc_data=pd.DataFrame(d)
acc_data

Number of trees  Accuracy
0              100  0.794598
1              200  0.787491
2              300  0.792466
3              400  0.792466
4              500  0.789623
5              600  0.791045
6              700  0.789623
7              800  0.788913
8              900  0.793177
9             1000  0.788202

From above dataframe we can observe that **600** trees will give the highest accuracy of **79%**.

##Conclusion
With this accuracy we can say that close to 80%(79.1%) of predictions gone through this random forest will be correct for newly added customers

##Including out of bag error values

In [82]:
from sklearn.ensemble import RandomForestClassifier
classifier1 = RandomForestClassifier(n_estimators=100, oob_score=True)#n estimator is num of trees
classifier1.fit(x_train,y_train)
oob_error = 1 - classifier1.oob_score_
oob_error

0.20924444444444446

In [81]:
from sklearn.ensemble import RandomForestClassifier

classifier1=RandomForestClassifier(n_estimators=300,oob_score=True)
classifier1.fit(x_train,y_train)

y_pred=classifier1.predict(x_test)

from sklearn.metrics import accuracy_score
accuracy_score(y_pred,y_test)

0.7882018479033405

In [85]:
# Specify the target column and the feature columns
target_column = df["Churn"]
feature_columns = df.iloc[:,:-1] # features

# Extract the target variable
y = df["Churn"]
X = df.iloc[:,:-1]
# One-hot encode categorical features
# X = pd.get_dummies(churn_data[feature_columns])

# Create a random forest classifier with a desired number of trees (n_estimators)
rf_classifier = RandomForestClassifier(n_estimators=100, oob_score=True, random_state=42)

# Fit the random forest on the data
rf_classifier.fit(X, y)

# Access the OOB dataset indices
oob_dataset_indices = np.where(rf_classifier.oob_decision_function_[:, 1] >= 0.5)

# Get the actual OOB dataset
oob_dataset = X.iloc[oob_dataset_indices]

# Print the OOB dataset
print("Out-of-Bag Dataset:")
oob_dataset

Out-of-Bag Dataset:


gender  SeniorCitizen  Partner  Dependents  tenure  PhoneService  \
0          0              0        1           0       1             0   
2          1              0        0           0       2             1   
4          0              0        0           0       2             1   
5          0              0        0           0       8             1   
8          0              0        1           0      28             1   
...      ...            ...      ...         ...     ...           ...   
7013       0              0        0           0      40             1   
7018       1              0        1           1       1             1   
7029       0              1        0           0       6             0   
7032       1              1        0           0       1             1   
7041       1              1        1           0       4             1   

      MultipleLines  InternetService  OnlineSecurity  OnlineBackup  \
0                 1                0               0             2   
2                 0                0               2             2   
4                 0                1               0             0   
5                 2                1               0             0   
8                 2                1               0             0   
...             ...              ...             ...           ...   
7013              2                1               0             2   
7018              0                1               0             0   
7029              1                0               0             0   
7032              2                1               0             0   
7041              2                1               0             0   

      DeviceProtection  TechSupport  StreamingTV  StreamingMovies  Contract  \
0                    0            0            0                0         0   
2                    0            0            0                0         0   
4                    0            0            0                0         0   
5                    2            0            2                2         0   
8                    2            2            2                2         0   
...                ...          ...          ...              ...       ...   
7013                 2            0            2                0         0   
7018                 0            0            0                0         0   
7029                 0            0            2                2         0   
7032                 0            0            0                0         0   
7041                 0            0            0                0         0   

      PaperlessBilling  PaymentMethod  MonthlyCharges  TotalCharges  
0                    1              2           29.85         29.85  
2                    1              3           53.85        108.15  
4                    1              2           70.70        151.65  
5                    1              2           99.65        820.50  
8                    1              2          104.80       3046.05  
...                ...            ...             ...           ...  
7013                 1              0           93.40       3756.40  
7018                 1              3           70.65         70.65  
7029                 1              2           44.40        263.05  
7032                 1              2           75.75         75.75  
7041                 1              3           74.40        306.60  

[1534 rows x 19 columns]